base/deepcopy.jl

```diff
@@ -124,7 +124,8 @@ end
     if haskey(stackdict, x)
         return stackdict[x]::typeof(x)
     end
-    stackdict[x] = $(Expr(:new, :(Array{T, N}), :(deepcopy_internal(x.ref, stackdict)), :(x.size)))
+    copied = $(Expr(:new, :(Array{T, N}), :(deepcopy_internal(x.ref, stackdict)), :(x.size)))
+    get!(stackdict, x, copied)::typeof(x)
```

Comment:
I think I'd put in some short comment here -- sure there's always git blame, but the double lookup may otherwise seem wrong to the naive reader.